src/EventStore.SourceGenerators/Messaging/StatsMessageSourceGenerator.cs

```diff
@@ -0,0 +1,279 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+
+// for speed and simplicity this does all its generation based on the syntax tree and not on the semantic model
+// there are essentially three phases
+// 1. use a SyntaxReceiver to detect and record uses of the attributes that we will generate code for
+// 2. Register each usage, discovering how they are related to each other in the abstract syntax tree bottom-up
+// 3. Process the resulting tree top-down, adding in the extra generated code that we want.
+namespace EventStore.SourceGenerators.Messaging {
+	[Generator]
+	public class StatsMessageSourceGenerator : ISourceGenerator {
```

Comment:
The generator should use [incremental version](https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.md).

The `ISourceGenerator` runs -- put simply -- on every keystroke, thus can slow down dev-experience on medium / large projects significantly. The incremental generator is more advanced, so that only on actual relevant change it's active. IMO also the programming model is more natural and aligned to how compilers work (parse and emit is more separated).

Reply:
thanks i'll take a look